TESTS/mbedmicro-rtos-mbed/signals/main.cpp

```diff
@@ -15,49 +15,357 @@
+void run_clear(void)
+{
+    ret_val = Thread::signal_clr(signals);
+    sem_a.release();
+}
+
+template <int32_t signals>
+void run_wiat_clear(void)
+{
+    sem_a.release();
+    sem_b.wait();
+    ret_val = Thread::signal_clr(signals);
+    sem_a.release();
+}
+
+template <int32_t signals1, int32_t signals2>
+void run_double_wiat_clear(void)
+{
+    sem_a.release();
+    sem_b.wait();
+    ret_val = Thread::signal_clr(signals1);
+    sem_a.release();
+    sem_b.wait();
+    ret_val = Thread::signal_clr(signals2);
+    sem_a.release();
+}
+
+void run_loop_wiat_clear(void)
+{
+    for (int i = 0; i <= MAX_FLAG_POS; i++) {
+        sem_b.wait();
+        ret_val = Thread::signal_clr(0);
+        sem_a.release();
+    }
+}
+
+
+/** Test signal_clr with no signals specified
+
+    Given two threads A and B running, B with all signals already set
+    When execute @a signal_clr(0) with no signals specified
+    Then thread B @a signal_clr status should be ALL_SIGNALS indicating that thread B state is unchanged
+ */
+void test_clear_0(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_double_wiat_clear<0, 0>));
+
+    sem_a.wait();
+    ret = tb.signal_set(ALL_SIGNALS);
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret);
+
+    sem_b.release();
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret_val);
+
+    sem_b.release();
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret_val);
+}
+
+/** Test if any signals are set on just created thread
+
+    Given one thread running
+    When execute @a signal_clr(0) with no signals specified
+    Then @a signal_clr return status should be 0 indicating no signals set
+ */
+void test_init_state(void)
+{
+    int32_t ret;
+    Thread t(osPriorityNormal, TEST_STACK_SIZE);
+    t.start(callback(run_clear<0>));
+
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(0, ret_val);
+}
+
+/** Test signal_set with all signals set in one shot
+
+    Given two threads A and B running
+    When thread A executes @a signal_set(ALL_SIGNALS) with all possible signals
+    Then thread B @a signal_clr(0) status should be ALL_SIGNALS indicating a all signals set correctly
+ */
+void test_set_all(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_wiat_clear<0>));
+
+    sem_a.wait();
+    ret = tb.signal_set(ALL_SIGNALS);
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret);
+
+    sem_b.release();
+    sem_a.wait();
+    TEST_ASSERT_EQUAL(ALL_SIGNALS, ret_val);
+}
+
+/** Test signal_set with prohibited signal
+
+    Given two threads A and B running, B with all signals set
+    When thread A executes @a signal_set(PROHIBITED_SIGNAL) with prohibited signal
+    Then thread B @a signal_clr(0) status should be ALL_SIGNALS indicating that thread B state is unchanged
+ */
+void test_set_prohibited(void)
+{
+    int32_t ret;
+    Thread tb(osPriorityNormal, TEST_STACK_SIZE);
+    tb.start(callback(run_wiat_clear<0>));
+
+    sem_a.wait();
```

Comment:
It's not in the description of the tests, do we need that here? Shouldn't we test separately that we can set signals?